### PR TITLE
planner: support specify table partition in optimize hint

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/17113

Problem Summary: support specify table parition in optimize hint

### What is changed and how it works?

What's Changed: support specify table parition in optimize hint

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`: after this PR merge, we should update the docs.
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- support specify table partition in optimize hint

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -141,18 +143,33 @@ func (tr *QueryTimeRange) Condition() string {
...
+		}
 		if tableInfo.dbName.L == "" {
 			tableInfo.dbName = defaultDBName
 		}
-		hintTableInfos[i] = tableInfo
+		switch hintName {
+		case TiDBMergeJoin, HintSMJ, TiDBIndexNestedLoopJoin, HintINLJ, HintINLHJ, HintINLMJ, TiDBHashJoin, HintHJ:
+			if len(tableInfo.partitions) > 0 {
+				ctx.GetSessionVars().StmtCtx.AppendWarning(
+					errors.New(fmt.Sprintf("%s hint do not support specify partitions", hintName)))
```
[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply planner/core/planbuilder.go:166]: The error message can be refined to `Optimizer Hint %s is inapplicable`

File: planner/core/testdata/integration_suite_in.json [hunk 2]
```diff
@@ -136,5 +136,18 @@
...
+    "name": "TestOptimizeHintOnPartitionTable",
+    "cases": [
+      "select /*+ use_index(t) */ * from t",
+      "select /*+ use_index(t partition(p0, p1) b, c) */ * from t",
+      "select /*+ ignore_index(t partition(p0, p1) b, c) */ * from t",
+      "select /*+ hash_join(t1, t2 partition(p0)) */ * from t t1 join t t2 on t1.a = t2.a",
+      "select /*+ use_index_merge(t partition(p0)) */ * from t where t.b = 1 or t.c = \"8\"",
+      "select /*+ use_index_merge(t partition(p0, p1) primary, b) */ * from t where t.a = 1 or t.b = 2",
+      "select /*+ use_index(t partition(p0) b) */ * from t partition(p0, p1)",
+      "select /*+ read_from_storage(tikv[t partition(p0)], tiflash[t partition(p1, p2)]) */ * from t"
```
Comment (XuHuaiyu) [inline planner/core/testdata/integration_suite_in.json:150]: Add test cases:
`select /*+ use_index(t) */ from t`
`select /*+ use_index(t partition(p0, p1) b, c) */ from t partition(p1,p2)`
`select /*+ use_index(t partition(p_non_exist)) */ from t partition(p1,p2)`

**Comment by lzmhhh123** [comment]
> Do we need to raise a warning for sql like if the partition is unknown in the hint:
> `select /*+ use_index(t partition(p_non_exist)) */ from t partition(p1,p2)`

OK, I try to support it.

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -210,16 +248,32 @@ func (info *tableHintInfo) matchTableName(tables []*hintTableInfo, hintTables []
...
-	buffer.WriteString(strings.ToUpper(hintType))
-	buffer.WriteString("(")
+func restore2TableHint(hintTables []hintTableInfo) string {
+	buffer := bytes.NewBufferString("")
 	for i, table := range hintTables {
 		buffer.WriteString(table.tblName.L)
+		if len(table.partitions) > 0 {
+			buffer.WriteString(" PARTITION(")
+			for j, partition := range table.partitions {
+				buffer.WriteString(partition.L)
```
Comment (XuHuaiyu) [inline planner/core/planbuilder.go:258]: This should be moved to line 262

File: planner/core/rule_partition_processor.go [hunk 4]
```diff
@@ -694,8 +696,110 @@ func pruneUseBinarySearch(lessThan lessThanDataInt, data dataForPrune, unsigned
...
+	}
+
+	return s.resolveAccessPaths(ds)
+}
+
+func checkTableHintsApplicableForPartition(partitions []model.CIStr, ds *DataSource, partitionSet set.StringSet) {
+	for _, p := range partitions {
+		if !partitionSet.Exist(p.L) {
+			ds.ctx.GetSessionVars().StmtCtx.AppendWarning(
+				errors.New(fmt.Sprintf("partition `%s` in hint can't hit any source", p.L)))
```
Comment (XuHuaiyu) [inline planner/core/rule_partition_processor.go:784]: Can we return the partition name back to `checkHintsApplicable`, and build a more comprehensible error message like `Unknow partition %s in optimizer hint %s`

**Comment by tiancaiamao** [comment]
LGTM

**Comment by tiancaiamao** [comment]
@XuHuaiyu

**Pull request merged**